nodeJS/express/views.md

```diff
@@ -0,0 +1,165 @@
+app.set("view engine", "ejs");
+```
+
+This enables `EJS` as the view engine, and that our app should look for templates in the `/views` subdirectory.
+
+### EJS syntax
+
+In `EJS`, the `<%` and `%>` tags allow us to use JavaScript. Whatever is in these tags will behave like JavaScript. This lets us write conditional statements, `for` loops, as well as use variables.
+
+In order to work with variables, we use the `<%=` tag. This lets us convert variables into values when the application is loaded. Otherwise, stick with `<%` when using JavaScript logic.
+
+Here's a quick example that makes includes arrays and loop logic.
+
+```js
+<% const animals = ["Cat", "Dog", "Lemur", "Hawk"] %>
+
+<% animals.map(animal => { %>
+- <%= animal%>s are cute <% }) %>
+```
+
+### Using EJS with Express
+
+Let's use `EJS` with `Express`. First, create an `EJS` template file called `index.ejs`, and add the following:
+
+```html
+<html>
+  <body>
+    <%= message %>
+  </body>
+</html>
+```
+
+And in app.js, render this template file in one of your routes:
+
+```js
+app.get("/", (req, res) => {
+  res.render("index", { message: "EJS rocks!" });
+});
+```
+
+When the `"/"` route is matched, the template file matching `"index"` is rendered with the message value passed, such that you should see:
+
+```
+EJS rocks!
+```
+
+### Reusable web components
+
+You may want to include webpage components that are shared across different pages, such as a sidebar or a header. To insert such components into your pages, we make use of the `include` command. This requires the name of the file to be inserted, and optionally an object of data you wish to pass.
+
+Say you have the following navbar component called `"navbar.ejs"`:
+
+```html
+<!-- navbar.ejs -->
+<nav>
+  <ul>
+    <% for (let i = 0; i < links.length; i++) { %>
+    <li>
+      <a href="<%= links[i].href %>">
+        <span> <%= links[i].text %> </span>
+      </a>
+    </li>
+    <% } %>
+  </ul>
+</nav>
+```
+
+You can insert this component into another EJS file like so:
+
+```html
+<!-- index.ejs -->
+<html>
+  <head>
+    <title>Homepage</title>
+    <body>
+      <%- include('navbar', {links: links}) %>
+    </body>
+  </head>
+</html>
+```
+
+This can be used to include headers and footers in all of your pages, for example.
+
+Note that the navbar expects a `links` value. To pass this data into the navbar, you can pass it when rendering `index.ejs` which contains the navbar. Modify `app.js` such that a `links` object is defined and passed into the `render` function in the `"/"` route handler:
+
+```js
+// app.js
+const links = [
+  { href: "/", text: "Home" },
+  { href: "about", text: "About" },
+];
+
+app.get("/", (req, res) => {
+  res.render("index", { links: links });
+});
+```
+
+Here's another example of how to use `includes` to dynamically render a list of variables:
+
+```html
+<ul>
+  <% users.map( user => { %> <%- include('user/show', {user: user}) %> <% }) %>
+</ul>
+```
+
```

Comment:
I'm not if learners will know what `user/show` will match since we haven't discussed that there can be subdirectories within the `views` folder. A lesson note would be sufficient.

```suggestion
<div class="lesson-note lesson-note--tip" markdown="1">

#### Directories within the views folder

We can have nested directories of EJS template files within the `views`. For example, to render the template file `./views/user/show.ejs`, we'll need to provide the relative path like so:

```js
// in res.render
res.render("user/show")

// in include
include("user/show")
```

</div>
```

Reply:
That makes sense that the learner won't be aware of what user/show is, thank you for catching my oversight.